hypothesis-python/src/hypothesis/strategies/_internal/numbers.py

```diff
@@ -182,97 +182,105 @@ def integers(
-    def __init__(self, allow_infinity, allow_nan, allow_subnormal, width):
+class FloatStrategy(SearchStrategy):
+    """A strategy for floating point numbers."""
+
+    def __init__(
+        self,
+        min_value: float = -math.inf,
+        max_value: float = math.inf,
+        allow_nan: bool = True,
+        # The smallest nonzero number we can represent is usually a subnormal, but may
+        # be the smallest normal if we're running in unsafe denormals-are-zero mode.
+        # While that's usually an explicit error, we do need to handle the case where
+        # the user passes allow_subnormal=False.
+        smallest_nonzero_magnitude: float = SMALLEST_SUBNORMAL,
+    ):
         super().__init__()
-        assert isinstance(allow_infinity, bool)
         assert isinstance(allow_nan, bool)
-        assert isinstance(allow_subnormal, bool)
-        assert width in (16, 32, 64)
-        self.allow_infinity = allow_infinity
+        assert smallest_nonzero_magnitude >= 0.0
+        self.min_value = min_value
+        self.max_value = max_value
         self.allow_nan = allow_nan
-        self.allow_subnormal = allow_subnormal
-        self.width = width
-
+        self.smallest_nonzero_magnitude = smallest_nonzero_magnitude
+
+        boundary_values = [
+            min_value,
+            next_up(min_value),
+            min_value + 1,
+            max_value - 1,
+            next_down(max_value),
+            max_value,
+        ]
         self.nasty_floats = [
-            float_of(f, self.width) for f in NASTY_FLOATS if self.permitted(f)
+            f for f in NASTY_FLOATS + boundary_values if self.permitted(f)
         ]
         weights = [0.2 * len(self.nasty_floats)] + [0.8] * len(self.nasty_floats)
-        self.sampler = d.Sampler(weights)
+        self.sampler = d.Sampler(weights) if self.nasty_floats else None
+
+        self.pos_clamper = self.neg_clamper = None
+        if _sign_aware_lte(0.0, max_value):
+            pos_min = max(min_value, smallest_nonzero_magnitude)
+            allow_zero = _sign_aware_lte(min_value, 0.0)
+            self.pos_clamper = make_float_clamper(pos_min, max_value, allow_zero)
+        if _sign_aware_lte(min_value, -0.0):
+            neg_max = min(max_value, -smallest_nonzero_magnitude)
+            allow_zero = _sign_aware_lte(-0.0, max_value)
+            self.neg_clamper = make_float_clamper(-neg_max, -min_value, allow_zero)
+
+        self.forced_sign_bit: Optional[int] = None
+        if (self.pos_clamper is None) != (self.neg_clamper is None):
+            self.forced_sign_bit = 1 if self.neg_clamper else 0
 
     def __repr__(self):
-        return "{}(allow_infinity={}, allow_nan={}, width={})".format(
-            self.__class__.__name__, self.allow_infinity, self.allow_nan, self.width
+        return "{}(min_value={}, max_value={}, allow_nan={})".format(
+            self.__class__.__name__,
+            self.min_value,
+            self.max_value,
+            self.allow_nan,
```

Comment:
Let's add `smallest_nonzero_magnitude` here too.

(optional: this might be easier to read if we make the *constructor_argument* `allow_subnormal: bool`, and then translate that to a float attribute `self.smallest_nonzero_magnitude`?)

Reply:
Ah yes, done